stl/inc/flat_set

```diff
@@ -165,7 +166,9 @@ public:
         _Guard._Target = nullptr;
         return *this;
     }
-    _Base_flat_set& operator=(_Base_flat_set&& _Other) {
+    _Base_flat_set& operator=(_Base_flat_set&& _Other) noexcept(
+        is_nothrow_move_assignable_v<container_type>&& is_nothrow_copy_assignable_v<key_compare>) // strengthened

```

Comment:
> This PR is currently making move functions unconditionally noexcept(false), which looks more buggy to me.

Yes, added. (I'm adding `// strengthened`; are these conditional noexcept mandated by the standard?)

> It seems that we can wrap the comparator into the following class to enforce copy on moving and leave the copy/move confuntions defaulted.

I think there is no need to introduce a copy wrapper, as they are solely for defaulting the methods, but the defaulted functions are not capable to deal with a lot of invariant-breaking cases.

Reply:
> are these conditional noexcept mandated by the standard?

I _think_ sometimes exposition only non-static data members are considered completely specifying the behavior of the implicitly declared or explicitly defaulted special member functions, e.g. for many `ranges` types and flat container adaptors, but definitely not for `optional`.

However, it's unfortunately underspecified whether non-static data members are considered complete. It seems permitted for implementations to add "evil" subobjects that weaken defaulted exception specifications.

Per [[functions.within.classes]](https://eel.is/c++draft/functions.within.classes), I think the standard currently implicitly requires that the copy/move operations are effectively defaulted and are able to break invariants.

We should report an LWG issue that makes copy/move functions explicitly specified. I believe we should say `// per an unnumbered LWG issue` instead of `// strengthened` for now.